### ISSUE TITLE
Distributing softwares that use Lambda-term

### ISSUE THREAD

**Issue by tmattio** [ocaml-community/lambda-term#93]
First off, thanks for the great work with Lambda-Term.
I recently created a [library](https://github.com/tmattio/inquire) that provides common CLI prompts, and it would have been 100x harder without LTerm.

I am trying to distribute a project that uses this library (and thus depends on Lambda-Term), and I'm having an issue because Camomile requires some files to be present in the file system.

This is fine when distributing on Opam, but I am also distributing on Homebrew and NPM, and I really don't want to package the 500 files Camomile depends on with the installer.

I realize the issue is more related to Camomile, but the way I see it, this dependency on local files is an inherent characteristic of Camomile that is unlikely to change.
As it has the unfortunate consequence of making applications that use Lambda-Term harder to distribute, I was wondering if there was something that could be done at this level to remove this pain-point.

From what I understand, Camomile is used by Zed_string. One solution I see is to functorize modules that use Zed_string to provide other string implementation, so users could use Lambda-Term with the standard String module for instance.

The distribution of my project is currently blocked by this so I can either re-implement the part that depends on Lambda-Term, or if you think a solution for the above is viable, I'm happy to give a shot at implementing it :)

EDIT:

To give additional context, this is the error I have when using a binary build on one machine on another:
```
Fatal error: exception Not_found
Raised at file "Camomile/internal/database.ml", line 50, characters 52-67
Called from file "Camomile/internal/unidata.ml" (inlined), line 187, characters 4-48
Called from file "Camomile/public/uCharInfo.ml", line 376, characters 16-46
Called from file "src/zed_utf8.ml", line 816, characters 13-53
```

I've tried to package the share files with my binary, but I haven't been able to run it when placing them in `/usr/local/share/camomile`, `/usr/share/camomile` or in the same directory as the binary.

**Comment by kandu** [comment]
Hi, @tmattio on the camomile issue, @nojb had already done a lot of fancy work. see #74 and ocaml-community/zed#16.

And the road blockes, that encoding conversion, the problematic windows console.. were discussed in #74.

Since @nojb seems to have no time to continually push this, so I just branched his work and built a working utop to confirm that the changes works properly.

The uu branches of zed, lambda-term, utop are published on my git-repo respectively, with no camomile dependency, so no filesystem files are needed. But as discussed in #74, utf8 encoding is now **mandatory**

https://github.com/kandu/zed/tree/uu
https://github.com/kandu/lambda-term/tree/uu
https://github.com/kandu/utop/tree/uu

**Comment by kandu** [comment]
Shall we make some releases named zed-uu, lambda-term-uu in my personal opam repo to make distributing more convenient? Or you want to maintain your own repo?

**Comment by kandu** [comment]
And to adapt lambda-term-uu
https://github.com/kandu/inquire/commit/2ecd2014d39f35cd05571e811df1baee9073671f

**Comment by Drup** [comment]
> Shall we make some releases named zed-uu, lambda-term-uu in my personal opam repo to make distributing more convenient? Or you want to maintain your own repo?

No, please make this part of lambda-term proper through a new major version (and take the opportunity to fix all the nits that were introduced in the zed API recently :) )

**Comment by tmattio** [comment]
Thanks a lot for looking into this and making the changes in Inquire @kandu!

I'll wait for the next version of lambda-term to be released (or vendor your repositories if I need it sooner). In the meantime, let me know if I can be of any assistance 🙂

**Comment by jonahbeckford** [comment]
Per @kandu :
> The uu branches of zed, lambda-term, utop are published on my git-repo respectively, with no camomile dependence, so no filesystem files are needed. But as discussed in #74, utf8 encoding is now **mandatory**
> 
> https://github.com/kandu/zed/tree/uu
> https://github.com/kandu/lambda-term/tree/uu
> https://github.com/kandu/utop/tree/uu

Since this issue is blocked on zed, I rebased @kandu 's changes with the latest from master, and added a few more commits to a) get mouse support in Windows and b) fix the display of multiwidth Unicode characters on Windows.

https://github.com/diskuv/zed/tree/uu
https://github.com/diskuv/lambda-term/tree/uu

I've tested on Windows with all the examples and compared it to the output on macOS. Only problems are:
a) I have no idea what "focus.ml" is supposed to do (but it behaves the same as macOS), and
b) "move.ml" does not have a square boundary on Windows because Zed/Uucp calculates the terminal width of "m̀é" as 4 characters (but Uucp is clear that it is not a 100% accurate terminal width "hint" so that is fine).
